core/src/main/java/org/apache/iceberg/rest/CatalogHandlers.java

```diff
@@ -0,0 +1,315 @@
+    UnprocessableEntityException(Set<String> commonKeys) {
+      super(String.format("Invalid namespace update, cannot set and remove keys: %s", commonKeys));
+    }
+  }
+
+  private static class ValidationFailureException extends RuntimeException {
+    private final CommitFailedException wrapped;
+
+    private ValidationFailureException(CommitFailedException cause) {
+      super(cause);
+      this.wrapped = cause;
+    }
+
+    public CommitFailedException wrapped() {
+      return wrapped;
+    }
+  }
+
+  public static ListNamespacesResponse listNamespaces(SupportsNamespaces catalog, Namespace parent) {
+    List<Namespace> results;
+    if (parent.isEmpty()) {
+      results = catalog.listNamespaces();
+    } else {
+      results = catalog.listNamespaces(parent);
+    }
+
+    return ListNamespacesResponse.builder().addAll(results).build();
+  }
+
+  public static CreateNamespaceResponse createNamespace(SupportsNamespaces catalog, CreateNamespaceRequest request) {
+    Namespace namespace = request.namespace();
+    catalog.createNamespace(namespace, request.properties());
+    return CreateNamespaceResponse.builder()
+        .withNamespace(namespace)
+        .setProperties(catalog.loadNamespaceMetadata(namespace))
+        .build();
+  }
+
+  public static GetNamespaceResponse loadNamespace(SupportsNamespaces catalog, Namespace namespace) {
+    Map<String, String> properties = catalog.loadNamespaceMetadata(namespace);
+    return GetNamespaceResponse.builder()
+        .withNamespace(namespace)
+        .setProperties(properties)
+        .build();
+  }
+
+  public static DropNamespaceResponse dropNamespace(SupportsNamespaces catalog, Namespace namespace) {
+    boolean dropped = catalog.dropNamespace(namespace);
+    return DropNamespaceResponse.builder()
+        .dropped(dropped)
+        .build();
+  }
+
+  public static UpdateNamespacePropertiesResponse updateNamespaceProperties(
+      SupportsNamespaces catalog, Namespace namespace, UpdateNamespacePropertiesRequest request) {
+    Set<String> removals = Sets.newHashSet(request.removals());
+    Map<String, String> updates = request.updates();
+
+    Set<String> commonKeys = Sets.intersection(updates.keySet(), removals);
+    if (!commonKeys.isEmpty()) {
+      throw new UnprocessableEntityException(commonKeys);
+    }
+
+    Map<String, String> startProperties = catalog.loadNamespaceMetadata(namespace);
+    Set<String> missing = Sets.difference(removals, startProperties.keySet());
+
+    if (!updates.isEmpty()) {
+      catalog.setProperties(namespace, updates);
+    }
+
+    if (!removals.isEmpty()) {
+      // remove the original set just in case there was an update just after loading properties
+      catalog.removeProperties(namespace, removals);
+    }
+
+    return UpdateNamespacePropertiesResponse.builder()
+            .addMissing(missing)
+            .addUpdated(updates.keySet())
+            .addRemoved(Sets.difference(removals, missing))
```

Comment:
This may not be true at this point since (as your comment alludes to above) we may have removed properties that were set in between our determination of missing and our execution of catalog.removeProperties. Not sure if we can make a better response though unless we change catalog.removeProperties to return which properties actually got removed.

Reply:
This is something for implementations to improve on. Right now there's no way to do this through the catalog API.